File: csharp/Ice/Middleware/Server/AuthorizationMiddleware.cs
// Copyright (c) ZeroC, Inc.

namespace Server;

/// <summary>The AuthorizationMiddleware intercepts incoming requests and checks the token entry in their contexts.
/// </summary>
internal class AuthorizationMiddleware : Ice.Object
{
    private readonly Ice.Object _next;
    private readonly string _validToken;

    /// <inheritdoc />
    public ValueTask<Ice.OutgoingResponse> dispatchAsync(Ice.IncomingRequest request)
    {
        // Check if the request has a valid token in its context.
        if (!request.current.ctx.TryGetValue("token", out string? token) || token != _validToken)
        {
            Console.WriteLine($"Rejecting request with invalid token '{token}'");
            throw new Ice.DispatchException(Ice.ReplyStatus.Unauthorized, $"Invalid token '{token}'");
        }

        Console.WriteLine($"Accepting request with token '{token}'");

        // Continue the dispatch pipeline.
        return _next.dispatchAsync(request);
    }

    /// <summary>Constructs an AuthorizationMiddleware.</summary>
    /// <param name="next">The next Ice.Object in the dispatch pipeline.</param>
    /// <param name="validToken">The value to check against.</param>
    internal AuthorizationMiddleware(Ice.Object next, string validToken)
    {
        _next = next;
        _validToken = validToken;
    }
}

File: java/Ice/middleware/server/src/main/java/com/example/ice/middleware/server/AuthorizationMiddleware.java
// Copyright (c) ZeroC, Inc.

package com.example.ice.middleware.server;

import com.zeroc.Ice.DispatchException;
import com.zeroc.Ice.IncomingRequest;
import com.zeroc.Ice.OutgoingResponse;
import com.zeroc.Ice.ReplyStatus;
import com.zeroc.Ice.UserException;
import java.util.concurrent.CompletionStage;

/**
 * The AuthorizationMiddleware intercepts incoming requests and checks the token entry in their contexts.
 */
class AuthorizationMiddleware implements com.zeroc.Ice.Object {
    private final com.zeroc.Ice.Object _next;
    private final String _validToken;

    public CompletionStage<OutgoingResponse> dispatch(IncomingRequest request) throws UserException {
        String token = request.current.ctx.get("token");

        // Check if the request has a valid token in its context.
        if (token == null || !token.equals(_validToken)) {
            System.out.println("Rejecting request with invalid token " + token + "'");
            throw new DispatchException(ReplyStatus.Unauthorized.value(), "Invalid token '" + token + "'");
        }

        System.out.println("Accepting request with token '" + token + "'");

        // Continue the dispatch pipeline.
        return _next.dispatch(request);
    }

    /**
     * Creates an AuthorizationMiddleware.
     *
     * @param next the next Ice.Object in the dispatch pipeline
     * @param validToken the value to check against
     */
    AuthorizationMiddleware(com.zeroc.Ice.Object next, String validToken) {
        _next = next;
        _validToken = validToken;
    }
}
